### PR TITLE
fix: accumulate float16 reductions in float32

### LINKED ISSUE #123299: tf.reduce_mean returns an inaccurate result for a constant float16 tensor, causing tf.math.reduce_std to produce a nonzero standard deviation
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

v2.21.0-rc1-5-ga481b10260d 2.21.0

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.reduce_mean` returns an inaccurate result for a constant float16 tensor containing large values. As a consequence, `tf.math.reduce_std` returns a nonzero standard deviation even though all input elements are identical.
For a constant tensor where every element equals `10000.0`, the expected mean is exactly `10000.0` and the standard deviation should be exactly `0.0`. However, TensorFlow computes a mean of approximately `9.99e+03 (i.e., 9992)`, which subsequently causes `tf.math.reduce_std` to return `8.0`.
Since all elements in the input tensor have the same value, the standard deviation is expected to be exactly 0.

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf 
x_np = np.array( 
    [ 
        [10000.0, 10000.0], 
        [10000.0, 10000.0], 
        [10000.0, 10000.0], 
    ], 
    dtype=np.float16, 
) 
x = tf.constant(x_np, dtype=tf.float16)
mean = tf.reduce_mean(x) 
std = tf.math.reduce_std(x) 
print("TensorFlow mean:", mean) 
print("TensorFlow std :", std) 
print("NumPy mean:", np.mean(x_np, dtype=np.float16))
print("NumPy std :", np.std(x_np, dtype=np.float16))
```

### Relevant log output

```shell
TensorFlow mean: tf.Tensor(9.99e+03, shape=(), dtype=float16)
TensorFlow std : tf.Tensor(8.0, shape=(), dtype=float16)
NumPy mean: 1e+04
NumPy std : 0.0
```

### PR DESCRIPTION
**Pull request by Ashutosh0x**
### Summary

`reduction_ops.h` gives `bfloat16` a `float` accumulator for both `SumReducer` and `MeanReducer`, under the comment *"Specialization for BF16 Reducer to fix accuracy"*. `Eigen::half` is simply missing from both lists, so `float16` reductions still accumulate in `float16` and lose precision long before the result type does.

Fixes #123299.

### Reproduction

Verified on `tf-nightly 2.22.0-dev20260726`:

```python
import numpy as np, tensorflow as tf
x = tf.constant(np.full((3, 2), 10000.0, dtype=np.float16))

tf.reduce_sum(x)       # 5.997e+04   -- expected 60000
tf.reduce_mean(x)      # 9.99e+03    -- expected 10000
tf.math.reduce_std(x)  # 8.0         -- expected 0.0, every element is identical
```

Both 60000.0 and 10000.0 are *exactly* representable in `float16` (max is 65504), so this is purely an accumulator-width problem, not an unavoidable rounding limit. The `float32` path returns `10000.0` and `0.0`, and NumPy also returns `1e+04` / `0.0` because it likewise accumulates `float16` in `float32`.

### Fix

Two lines, mirroring what `bfloat16` already does:

```c++
CASTING_SPECIALIZATION(Eigen::internal::SumReducer, Eigen::half, float);
...
CASTING_SPECIALIZATION(Eigen::half, float);   // MeanReducer
```

I also reworded the comment above the first macro, since it now covers more than BF16.

### Tests

* `SumReductionTest.testFloat16LargeValues` — the sum of six `10000.0` elements is exactly `60000`.
* `MeanReductionTest.testFloat16ConstantInput` — the mean of a constant tensor is exact and `reduce_std` is `0`.

Both are written against the existing `testBfloat16` / `testBFloat16` neighbours. A generic `_makeIncremental` sweep would not catch this, since the values there are small enough that `float16` accumulation is already exact.

### Two things worth a reviewer's judgement

1. **Performance.** This makes CPU `float16` reductions cast to `float`, exactly as `bfloat16` already does. I assume that trade was considered acceptable when the `bfloat16` specialization landed, but `float16` may have different callers.
2. **Verification.** I could not build TensorFlow on this machine, so the *bug* is verified against nightly but the *fix* is not runtime-verified, and I have not been able to run the wider reduction test suite. Results become strictly more accurate, so tolerance-based tests should be unaffected, but that is reasoning rather than a test run.

### CODE REVIEW

**Review by sgarciagoogle — CHANGES_REQUESTED** [review]
Thank you for this fix! However, the PR is not yet ready. Please address the following items:
  
  1. **Remove changes to `RELEASE.md`:**
     Please completely remove `RELEASE.md` from this PR. Your branch currently introduces +11,677 / -11,675 lines of churn due to line-ending conversions across the entire file. Release notes are managed separately, and including this file creates a massive diff that obscures reviewability and cannot be merged.
  
  2. **Use `assertAllClose` in unit tests:**
     In `tensorflow/python/kernel_tests/math_ops/reduction_ops_test.py` (in `testFloat16LargeValues` around line 283 and `testFloat16ConstantInput` around line 590), please replace `self.assertEqual(...)` with `self.assertAllClose(...)` as requested by the automated review bot. Exact equality assertions on floating-point reductions can cause test flakiness across different hardware targets and compilers.
  
Once done let us know and we will be happy to review again!

**Comment by Ashutosh0x** [comment]
@sgarciagoogle Thanks — both items are done in 5d51ce3.

**RELEASE.md removed.** The file is no longer touched by this PR at all. The churn was a line-ending conversion: my local copy was written back as CRLF while master's is LF, so every line in the file registered as changed. I rebuilt the branch on current master and re-applied only the two real files, so nothing about RELEASE.md — content or line endings — is carried in the diff now. That also clears the merge conflict, which was entirely in that file.

The diff is now **+26 / -2 across 2 files** (was +11,705 / -11,677 across 3):

```
 tensorflow/core/kernels/reduction_ops.h            |  9 +++++++--
 .../kernel_tests/math_ops/reduction_ops_test.py    | 19 +++++++++++++++++++
```

**assertAllClose.** Both tests now use it, as the bot suggested:

```python
    self.assertAllClose(tf_sum, 60000.0)
...
      self.assertAllClose(self.evaluate(math_ops.reduce_mean(x)), 10000.0)
      self.assertAllClose(self.evaluate(math_ops.reduce_std(x)), 0.0)
```

The kernel change in `reduction_ops.h` is byte-identical to what you reviewed — the two `CASTING_SPECIALIZATION` lines for `Eigen::half` and the reworded comment. Ready for another look whenever you have a moment.

One note carried over from the description, since it is still true and is the main thing worth a reviewer's judgement: I could not build TensorFlow locally, so the bug is verified against `tf-nightly 2.22.0-dev20260726` but the fix is not runtime-verified on my end. The CPU float16 reduction now casts to float exactly as bfloat16 already does, so if that performance trade needs a second opinion for FP16's callers, I am happy to follow whatever you decide.

**Review by sgarciagoogle — APPROVED** [review]
